python/ray/data/dataset.py

```diff
@@ -960,6 +961,123 @@ def add_column(batch: DataBatch) -> DataBatch:
+
+        .. warning::
+            ``distinct`` is an expensive operation that requires shuffling data across the cluster.
+            Large datasets may incur significant computation and memory cost.
+
+        Examples:
+
+            Remove duplicate rows across all columns:
+
+            .. testcode::
+
+                import ray
+                import pyarrow as pa
+
+                ds = ray.data.from_arrow(pa.table({"a": [1,2,1,2,3], "b": ["x","y","x","y","z"]}))
+                print(ds.distinct().sort(key=["a","b"]).take_all())
+
+            .. testoutput::
+
+                [{'a': 1, 'b': 'x'}, {'a': 2, 'b': 'y'}, {'a': 3, 'b': 'z'}]
+
+            Remove duplicates based on a subset of columns:
+
+            .. testcode::
+
+                print(ds.distinct(keys=["a"]).sort(key=["a"]).take_all())
+
+            .. testoutput::
+
+                [{'a': 1, 'b': 'x'}, {'a': 2, 'b': 'y'}, {'a': 3, 'b': 'z'}]
+
+            Get one record per group using a more complex example:
+
+            .. testcode::
+
+                ds3 = ray.data.from_arrow(pa.table({
+                    "id": [1, 1, 1, 2, 2],
+                    "timestamp": [100, 200, 300, 150, 250],
+                    "value": ["a", "b", "c", "d", "e"]
+                }))
+
+                # Get one record per id
+                unique_by_id = ds3.distinct(keys=["id"])
+                print(unique_by_id.sort(key=["id"]).take_all())
+
+            .. testoutput::
+
+                [{'id': 1, 'timestamp': 100, 'value': 'a'}, {'id': 2, 'timestamp': 150, 'value': 'd'}]
+
+        Args:
+            keys: List of columns to consider for identifying duplicates. Only the values in these columns are checked for duplication. If None, all columns are used. Defaults to None.
+
+        Returns:
+            A new dataset with duplicate rows removed, keeping one arbitrary row from each set of duplicates.
+        """
+        # Validate keys parameter.
+        if keys is not None:
+            if not keys:
+                raise ValueError(
+                    "keys cannot be an empty list. Use None to select all columns."
+                )
+
+            # Check for duplicate keys.
+            if len(keys) != len(set(keys)):
+                duplicates = [key for key in set(keys) if keys.count(key) > 1]
+                raise ValueError(
+                    f"Duplicate keys found: {duplicates}. Keys must be unique."
+                )
+
+            # Check that all specified keys exist in the dataset.
+            # Fetching the schema can trigger execution, so only do it when validating keys.
+            all_cols = self.columns()
+            if all_cols is not None:
+                invalid_keys = set(keys) - set(all_cols)
+                if invalid_keys:
+                    raise ValueError(
+                        f"Keys {list(invalid_keys)} not found in dataset columns {all_cols}"
+                    )
```

Comment:
Do this opportunistically to avoid triggering execution `schema(fetch_if_missing=False)`